src/dune_rules/gen_rules.ml

```diff
@@ -618,7 +618,10 @@ let gen_rules_standalone_or_root
     Rules.union rules rules'
   in
   let* build_config =
-    let+ directory_targets = collect_directory_targets ~dir ~init:directory_targets in
+    let+ directory_targets =
+      let+ directory_targets = collect_directory_targets ~dir ~init:directory_targets in
+      Gen_rules.Directory_targets.of_map directory_targets
```

Comment:
If the goal is to construct these things lazily why are we calling `collect_directory_targets` eagerly here? What is it exactly that's supposed to be lazy? Just the unions?